### PR TITLE
[Feature][ST-Engine] Add handle checkpoint timeout

### PR DESCRIPTION
**Pull request by EricJoy2048**

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java [hunk 1]
```diff
@@ -85,53 +102,74 @@ public PhysicalPlan(@NonNull List<SubPlan> pipelineList,
         }
         this.jobFullName = String.format("Job %s (%s)", jobImmutableInformation.getJobConfig().getName(),
             jobImmutableInformation.getJobId());
+
+        pipelineSchedulerFutureMap = new HashMap<>(pipelineList.size());
+    }
+
+    public void initJobMaster(JobMaster jobMaster) {
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java]: rename to `setJobMaster`?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java [hunk 2]
```diff
@@ -85,53 +102,74 @@ public PhysicalPlan(@NonNull List<SubPlan> pipelineList,
...
-        pipelineList.forEach(subPlan -> {
-            PassiveCompletableFuture<PipelineState> future = subPlan.initStateFuture();
-            future.whenComplete((v, t) -> {
-                // We need not handle t, Because we will not return t from Pipeline
+        pipelineList.forEach(subPlan -> addPipelineEndCallback(subPlan));
+    }
+
+    private void addPipelineEndCallback(SubPlan subPlan) {
+        PassiveCompletableFuture<PipelineState> future = subPlan.initStateFuture();
+        future.whenComplete((v, t) -> {
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java]: (v, t) -> (state, throwable)?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java [hunk 3]
```diff
@@ -85,53 +102,74 @@ public PhysicalPlan(@NonNull List<SubPlan> pipelineList,
...
                 if (PipelineState.CANCELED.equals(v)) {
+                    if (needRestore) {
+                        restorePipeline(subPlan);
+                        return;
+                    }
                     canceledPipelineNum.incrementAndGet();
+                    if (makeJobEndWhenPipelineEnded) {
+                        cancelJob();
+                    }
+                    jobMaster.releasePipelineResource(subPlan.getPipelineId());
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java:134]: Add log?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java [hunk 4]
```diff
@@ -164,6 +193,7 @@ public Address queryTaskGroupAddress(long taskGroupId) {
     }
 
     public void cancelJob() {
+        physicalPlan.neverNeedRestore();
         this.physicalPlan.cancelJob();
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java]: Add `this` or remove `this`?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalVertex.java [hunk 5]
```diff
@@ -229,25 +230,25 @@ private TaskGroupImmutableInformation getTaskGroupImmutableInformation() {
...
             String message = "Need a end state, not " + endState;
-            LOGGER.severe(message);
-            throw new IllegalStateException(message);
+            LOGGER.warning(message);
+            return false;
         }
 
         LOGGER.info(String.format("%s turn to end state %s.",
             taskFullName,
             endState));
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalVertex.java]: Add `taskFullName` to log message?

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java [hunk 6]
```diff
@@ -134,27 +142,48 @@ public void run() {
                 }
                 jobMasterCompleteFuture.complete(physicalPlan.getJobStatus());
             });
-            ownedSlotProfiles.putAll(new PipelineBaseScheduler(physicalPlan, this).startScheduling());
+            jobScheduler = new PipelineBaseScheduler(physicalPlan, this);
+            scheduleFuture = CompletableFuture.runAsync(() -> {
+                ownedSlotProfiles.putAll(jobScheduler.startScheduling());
+            }, executorService);
+            scheduleFuture.join();
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java:149]: Add log before blocking code

File: seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/TestUtils.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+        Action fake = new SourceAction<>(idGenerator.getNextId(), "fake", fakeSource,
+            Sets.newHashSet(new URL("file:///fake.jar")));
+        fake.setParallelism(3);
+        LogicalVertex fakeVertex = new LogicalVertex(fake.getId(), fake, 3);
+
+        ConsoleSink consoleSink = new ConsoleSink();
+        consoleSink.setSeaTunnelContext(SeaTunnelContext.getContext());
+        Action console = new SinkAction<>(idGenerator.getNextId(), "console", consoleSink,
+            Sets.newHashSet(new URL("file:///console.jar")));
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/TestUtils.java:52]: Is this jar path valid?

Comment (EricJoy2048) [reply seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/TestUtils.java:52]: > 

No, it ok.

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java [hunk 8]
```diff
@@ -85,53 +102,78 @@ public PhysicalPlan(@NonNull List<SubPlan> pipelineList,
...
-            future.whenComplete((v, t) -> {
-                // We need not handle t, Because we will not return t from Pipeline
-                if (PipelineState.CANCELED.equals(v)) {
+        pipelineList.forEach(subPlan -> addPipelineEndCallback(subPlan));
+    }
+
+    private void addPipelineEndCallback(SubPlan subPlan) {
+        PassiveCompletableFuture<PipelineState> future = subPlan.initStateFuture();
+        future.whenComplete((pipelineState, throwable) -> {
+            // We need not handle t, Because we will not return t from Pipeline
```
Comment (ashulin) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java]:
```suggestion
        future.thenAcceptAsync(pipelineState -> {
```
If you do not need to handle throwable inside, you can use other methods

**Pull request merged**